### ISSUE TITLE
Slow stats

### ISSUE THREAD

**Issue by kraih** [Grinnz/Minion-Backend-SQLite#18]
I was having some performance issues in a project and noticed that `stats` does not appear to use any indexes.
```
sqlite> explain query plan select count(case when state = 'inactive' and (expires is null or expires > datetime('now')) then 1 end) as inactive_jobs, count(case state when 'active' then 1 end) as active_jobs, count(case state when 'failed' then 1 end) as failed_jobs, count(case state when 'finished' then 1 end) as finished_jobs, count(case when state = 'inactive' and delayed > datetime('now') then 1 end) as delayed_jobs, (select count(*) from minion_locks where expires > datetime('now')) as active_locks, count(distinct case when state = 'active' then worker end) as active_workers, ifnull((select seq from sqlite_sequence where name = 'minion_jobs'), 0) as enqueued_jobs, (select count(*) from minion_workers) as inactive_workers, null as uptime from minion_jobs;
QUERY PLAN
|--USE TEMP B-TREE FOR count(DISTINCT)
|--SCAN minion_jobs
|--SCALAR SUBQUERY 1
|  `--SCAN minion_locks
|--SCALAR SUBQUERY 2
|  `--SCAN sqlite_sequence
`--SCALAR SUBQUERY 3
   `--SCAN minion_workers
```
Bet that can be optimised a little (probably a lot). After all `stats` is meant to be as fast as possible, because the Minion admin ui polls it over and over. (In my case it was Telegraf polling for stats to feed them to Grafana btw.)

**Comment by kraih** [comment]
This will probably require a second index on `state`. Not sure what negative side effects splitting up the existing compound index might have on other queries. 🤔

**Comment by kraih** [comment]
The problem is actually a little more complicated than i thought.
```
sqlite> explain query plan select count(*) from minion_jobs where state = 'inactive' and (expires is null or expires > datetime('now'));
QUERY PLAN
`--SEARCH minion_jobs USING INDEX minion_jobs_state_priority_id (state=?)

sqlite> explain query plan select count(case when state = 'inactive' and (expires is null or expires > datetime('now')) then 1 end) as inactive_jobs from minion_jobs;
QUERY PLAN
`--SCAN minion_jobs
```
So the index that already exists works fine, it's just the query where the planner doesn't want to use it.

**Comment by kraih** [comment]
And this is where things get funny.
```
sqlite> explain query plan select (select count(*) from minion_jobs where state = 'inactive' and (expires is null or expires > datetime('now'))) as inactive_jobs, (select count(*) from minion_jobs where state = 'active') as active_jobs, (select count(*) from minion_jobs where state = 'failed') as failed_jobs;
QUERY PLAN
|--SCAN CONSTANT ROW
|--SCALAR SUBQUERY 1
|  `--SEARCH minion_jobs USING INDEX minion_jobs_state_priority_id (state=?)
|--SCALAR SUBQUERY 2
|  `--SEARCH minion_jobs USING COVERING INDEX minion_jobs_state_priority_id (state=?)
`--SCALAR SUBQUERY 3
   `--SEARCH minion_jobs USING COVERING INDEX minion_jobs_state_priority_id (state=?)
```
Separate selects on the same table look more efficient.

**Comment by kraih** [comment]
This full stats query might do the trick.
```
sqlite> explain query plan select (select count(*) from minion_jobs where state = 'inactive' and (expires is null or expires > datetime('now'))) as inactive_jobs, (select count(*) from minion_jobs where state = 'active') as active_jobs, (select count(*) from minion_jobs where state = 'failed') as failed_jobs, (select count(*) from minion_jobs where state = 'finished') as finished_jobs, (select count(*) from minion_jobs where state = 'inactive' and delayed > datetime('now')) as delayed_jobs, (select count(*) from minion_locks where expires > datetime('now')) as active_locks, (select count(distinct(worker)) from minion_jobs where state = 'active') as active_workers, ifnull((select seq from sqlite_sequence where name = 'minion_jobs'), 0) as enqueued_jobs, (select count(*) from minion_workers) as inactive_workers, null as uptime;
QUERY PLAN
|--SCAN CONSTANT ROW
|--SCALAR SUBQUERY 1
|  `--SEARCH minion_jobs USING INDEX minion_jobs_state_priority_id (state=?)
|--SCALAR SUBQUERY 2
|  `--SEARCH minion_jobs USING COVERING INDEX minion_jobs_state_priority_id (state=?)
|--SCALAR SUBQUERY 3
|  `--SEARCH minion_jobs USING COVERING INDEX minion_jobs_state_priority_id (state=?)
|--SCALAR SUBQUERY 4
|  `--SEARCH minion_jobs USING COVERING INDEX minion_jobs_state_priority_id (state=?)
|--SCALAR SUBQUERY 5
|  `--SEARCH minion_jobs USING INDEX minion_jobs_state_priority_id (state=?)
|--SCALAR SUBQUERY 6
|  `--SCAN minion_locks
|--SCALAR SUBQUERY 7
|  |--USE TEMP B-TREE FOR count(DISTINCT)
|  `--SEARCH minion_jobs USING INDEX minion_jobs_state_priority_id (state=?)
|--SCALAR SUBQUERY 8
|  `--SCAN sqlite_sequence
`--SCALAR SUBQUERY 9
   `--SCAN minion_workers
```

**Comment by kraih** [comment]
Before:
```
sqlite> select count(case when state = 'inactive' and (expires is null or expires > datetime('now')) then 1 end) as inactive_jobs, count(case state when 'active' then 1 end) as active_jobs, count(case state when 'failed' then 1 end) as failed_jobs, count(case state when 'finished' then 1 end) as finished_jobs, count(case when state = 'inactive' and delayed > datetime('now') then 1 end) as delayed_jobs, (select count(*) from minion_locks where expires > datetime('now')) as active_locks, count(distinct case when state = 'active' then worker end) as active_workers, ifnull((select seq from sqlite_sequence where name = 'minion_jobs'), 0) as enqueued_jobs, (select count(*) from minion_workers) as inactive_workers, null as uptime from minion_jobs;
2|5|0|2311|0|5|1|6426|1|
Run Time: real 0.005 user 0.000637 sys 0.004320
```
After:
```
sqlite> select (select count(*) from minion_jobs where state = 'inactive' and (expires is null or expires > datetime('now'))) as inactive_jobs, (select count(*) from minion_jobs where state = 'active') as active_jobs, (select count(*) from minion_jobs where state = 'failed') as failed_jobs, (select count(*) from minion_jobs where state = 'finished') as finished_jobs, (select count(*) from minion_jobs where state = 'inactive' and delayed > datetime('now')) as delayed_jobs, (select count(*) from minion_locks where expires > datetime('now')) as active_locks, (select count(distinct(worker)) from minion_jobs where state = 'active') as active_workers, ifnull((select seq from sqlite_sequence where name = 'minion_jobs'), 0) as enqueued_jobs, (select count(*) from minion_workers) as inactive_workers, null as uptime;
2|5|0|2311|0|5|1|6426|1|
Run Time: real 0.000 user 0.000259 sys 0.000435
```
So even on a fairly small database with only 2400 jobs the separate selects with index use are faster.

**Comment by Grinnz** [comment]
Thanks, released as v5.0.6
